python/lsst/dc2gen3/patch_dc2.py

```diff
@@ -19,70 +23,210 @@
+    _break_hardlink(calib_registry_filename)
+
+    with closing(sqlite3.connect(calib_registry_filename)) as con:
+        for table in tables_to_update:
+            for old_filter_name in old_filter_names:
+                new_filter_name = _transform_filter_name(old_filter_name)
+                query = f"UPDATE {table} SET filter='{new_filter_name}' WHERE filter='{old_filter_name}';"
+                log.debug(f"Executing query on {calib_registry_filename}: {query}")
+                with con:
+                    con.execute(query)
+                    # Split the message into several lines, which
+                    # makes the code worse but the linter happy.
+                    log.info(
+                        f"Changed old filter name '{old_filter_name}'"
+                        + f" in table {table} of {calib_registry_filename}"
+                        + f" to '{new_filter_name}'"
+                    )
+
+
+def replace_root(filename, old_root, new_root):
+    """Replace the original gen2 root with the new one
+
+    Parameters:
+    filename : `str`
+        the path of the file in which to replace the root
+    old_root : `str`
+        the original fully qualified path of the gen2 repo root
+    new_root : `str`
+        the new fully qualified path of the gen2 repo root
+    """
+    log = lsst.log.Log.getLogger("convertRepo")
+
+    # Not attempting to deal with parsing the yaml, because
+    # that would require building a special purpose constructor
+    # and emitter for RepositoryCfg_v1, which would only
+    # make it more fragile.
+
+    with open(filename, "r") as f:
+        config_str = f.read()
+
+    new_config_str = config_str.replace(old_root, new_root)
+    if new_config_str == config_str:
+        log.info(f"No change in {filename}")
+    else:
+        _break_hardlink(filename)
+        with open(filename, "w") as f:
+            f.write(new_config_str)
+        log.info(f"{filename} updated")
 
 
 # classes
 
 # internal functions & classes
 
 
+def _break_hardlink(filename):
+    """Break a hardlink between a file and any others, if necessary
+
+    Parameters
+    ==========
+    filename : `str`
+        the path of the file
+    """
+    # If it isn't hardlinked to anything else, there is nothing to do
+    if os.stat(filename).st_nlink <= 1:
+        return
+
+    parent_dir = pathlib.Path(filename).parent.absolute()
+    with tempfile.TemporaryDirectory(dir=parent_dir) as temp_dir:
+        temp_file = os.path.join(temp_dir, "linked_file")
+        shutil.move(filename, temp_file)
+
+        try:
+            shutil.copy(temp_file, filename)
+            assert (
+                os.stat(filename).st_nlink == 1
+            ), f"Failed to break hardlink on {filename}"
+
+            # the linter does not like a bare except, but
+            # I really do want to clean up no matter what
+            # the exception was, so catch the base of
+            # all exceptions.
```

Comment:
This is a context where I'd just use a `# noqa: <some number>` to quiet the linter.  Better (IMO) than changing code to satisfy the linter unless the the change is an improvement in its own right.

Reply:
Bare except is deprecated in python 3 and should not ever be used in new code. Catching `BaseException` vs `Exception` is a decision though since you will be catching CTRL-C with `BaseException`.

I am surprised we are using hardlinks in gen2 -- I thought gen2 always tended to use softlinks.

Reply:
I don't think the original gen2 butler at IN2P3 uses hard links, but I've been using hardlinks of our primary copy to protect it while getting things working: it's big enough that copying it is a big deal, and also big enough that making a real copy is costly. So, while developing the patch_dc2 code, I use a "cp -rlL" to make a second "copy" right next to our copy on disk, and run against that copy, only breaking the links when I actually need to alter file contents. This save space while protecting our primary copy.
Could I have used soft links instead? Probably.  Hard links do have the advantage that we can follow the same procedure on the full repo when we are ready, test it, and when we are satisfied, remove the original tree without breaking any links.